### PR TITLE
Feature/homa lite improvement

### LINKED ISSUE #1223: homa lite improvements
- enforce a minimal ksm to mint
- have a cap on how much ksm can be accepted. allow council to update this cap

### LINKED ISSUE #1230: Cross-chain transfer in homa-lite
`homa-lite` should be able to send KSM from Karura to the sovereign sub-account on Kusama
- On users request to stake, make the cross-chain transfer. User would pay the transfer fee.
- The sovereign sub-account should be a configurable type in pallet `Config`.

#### Storage

The `RelayChainStashAccount` is not needed anymore.

#### Dispatchable calls

- Remove `RelayChainStashAccount` related calls and configs.
- Use `XcmTransfer` trait to do cross-chain transfer. Make the impl configurable as a type in pallet `Config`.

### PR DESCRIPTION
**Pull request by syan095**
Improved the Homa Lite module:
* Added a minimum requirement for request_mint
* Added a cap on how much staking currency can be used to mint
* Integrated the Homa lite module into Karura runtime
* Integrated Cross chain transfer Xcm transfer into the Homa Lite module
* Added benchmarking for Homa lite to Karura

closes #1230
closes #1223

### CODE REVIEW

File: modules/homa-lite/src/lib.rs [hunk 1]
```diff
@@ -41,13 +41,17 @@ use frame_support::{pallet_prelude::*, transactional, PalletId};
...
+use orml_traits::XcmTransfer;
 use primitives::{Balance, CurrencyId, EraIndex};
 use sp_runtime::{ArithmeticError, FixedPointNumber};
 use sp_std::prelude::*;
+use xcm::opaque::v0::{MultiLocation, Outcome};
 
 pub use module::*;
 pub use weights::WeightInfo;
 
+pub type XcmExecutionResult = sp_std::result::Result<Outcome, DispatchError>;
```
Comment (shaunxw) [inline modules/homa-lite/src/lib.rs:53]: Use the import from orml.

File: modules/homa-lite/src/lib.rs [hunk 2]
```diff
@@ -87,17 +91,32 @@ pub mod module {
 
 		/// Origin represented by the Root or Governance
 		type GovernanceOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The minimal amount of Staking currency to be locked
+		#[pallet::constant]
+		type MinimumMintThreshold: Get<Balance>;
+
+		/// The interface to Cross-chain transfer.
+		type CrossChainTransfer: XcmTransfer<Self::AccountId, Balance, CurrencyId>;
```
Comment (shaunxw) [inline modules/homa-lite/src/lib.rs:100]:
```suggestion
		type XcmTransfer: XcmTransferT<Self::AccountId, Balance, CurrencyId>;
```

File: modules/homa-lite/src/lib.rs [hunk 3]
```diff
@@ -87,17 +91,32 @@ pub mod module {
...
-		/// The relay chain's stash account have not been set.
-		RelayChainStashAccountNotSet,
 		/// The total issuance for the Staking currency must be more than zero.
 		InvalidStakedCurrencyTotalIssuance,
+		/// The mint amount is below the minimum threshold allowed.
+		MintAmountBelowMinimumThreshold,
+		/// The amount of Staking currency used has exceeded the cap allowed.
+		ExceededStakingCurrencyMintCap,
+		/// Error has occurred during Cross-chain transfer.
+		XCMTransferError,
```
Comment (shaunxw) [inline modules/homa-lite/src/lib.rs:119]:
```suggestion
		XcmTransferFailed,
```

File: modules/homa-lite/src/lib.rs [hunk 4]
```diff
@@ -87,17 +91,32 @@ pub mod module {
...
+		/// The minimal amount of Staking currency to be locked
+		#[pallet::constant]
+		type MinimumMintThreshold: Get<Balance>;
+
+		/// The interface to Cross-chain transfer.
+		type CrossChainTransfer: XcmTransfer<Self::AccountId, Balance, CurrencyId>;
+
+		/// The sovereign sub-account for where the staking currencies are sent to.
+		#[pallet::constant]
+		type XcmSovereignSubAccount: Get<MultiLocation>;
```
Comment (shaunxw) [inline modules/homa-lite/src/lib.rs:104]:
```suggestion
		type SovereignSubAccountLocation: Get<MultiLocation>;
```

File: modules/homa-lite/src/lib.rs [hunk 5]
```diff
@@ -138,11 +157,17 @@ pub mod module {
 	#[pallet::getter(fn current_batch)]
 	pub type CurrentBatch<T: Config> = StorageValue<_, EraIndex, ValueQuery>;
 
-	/// The account in which the staking currency goes into to be transferred to the Relay chain.
-	/// RelayChainStashAccount: value: stash_account: AccountId
+	/// The cap on the total amount of staking currency allowed to mint Liquid currency.
+	/// StakingCurrencyMintCap: value: mint_cap: Balance
+	#[pallet::storage]
+	#[pallet::getter(fn staking_currency_mint_cap)]
+	pub type StakingCurrencyMintCap<T: Config> = StorageValue<_, Balance, OptionQuery>;
```
Comment (shaunxw) [inline modules/homa-lite/src/lib.rs:164]: `ValueQuery` should be fine. The default value `zero` is reasonable in this context, users cannot mint if cap not set.

File: modules/homa-lite/src/lib.rs [hunk 6]
```diff
@@ -155,21 +180,51 @@ pub mod module {
...
+				who.clone(),
+				staking_currency,
+				amount,
+				T::XcmSovereignSubAccount::get(),
+				xcm_dest_weight,
+			)?;
+			match xcm_result {
+				Outcome::Complete(_) => Ok(()),
+				_ => Err(Error::<T>::XCMTransferError),
+			}?;
```
Comment (shaunxw) [inline modules/homa-lite/src/lib.rs:222]: The readability could be improved:

```suggestion
			ensure!(matches!(xcm_result, Outcome::Complete(_)), Error::<T>::XCMTransferError);
```

More like plain English :)

File: modules/homa-lite/src/lib.rs [hunk 7]
```diff
@@ -236,19 +292,19 @@ pub mod module {
...
+		#[pallet::weight(< T as Config >::WeightInfo::set_staking_currency_cap())]
 		#[transactional]
-		pub fn set_stash_account_id(origin: OriginFor<T>, new_account_id: T::AccountId) -> DispatchResult {
+		pub fn set_staking_currency_cap(origin: OriginFor<T>, new_cap: Balance) -> DispatchResult {
 			// This can only be called by Governance or ROOT.
 			T::GovernanceOrigin::ensure_origin(origin)?;
 
-			RelayChainStashAccount::<T>::put(new_account_id.clone());
-			Self::deposit_event(Event::<T>::RelayChainStashAccountUpdated(new_account_id));
+			StakingCurrencyMintCap::<T>::put(new_cap.clone());
```
Comment (shaunxw) [inline modules/homa-lite/src/lib.rs:306]: You won't need `clone()` here.

File: modules/homa-lite/src/tests.rs [hunk 8]
```diff
@@ -42,90 +42,98 @@ fn mock_initialize_token_works() {
...
 			System::events().iter().last().unwrap().event,
-			Event::HomaLite(crate::Event::RelayChainStashAccountUpdated(RELAY_CHAIN_STASH))
+			Event::HomaLite(crate::Event::MintRequested(current_batch, ALICE, amount))
 		);
 	});
 }
 
 #[test]
-fn request_mint_works() {
+fn can_request_mint_more_than_once_in_an_batch() {
```
Comment (shaunxw) [inline modules/homa-lite/src/tests.rs:66]:
```suggestion
fn can_request_mint_more_than_once_in_a_batch() {
```

File: runtime/karura/src/lib.rs [hunk 9]
```diff
@@ -1403,6 +1405,29 @@ impl cumulus_pallet_dmp_queue::Config for Runtime {
...
+		id: ParachainInfo::get().into_account(),
+	});
+}
+impl module_homa_lite::Config for Runtime {
+	type Event = Event;
+	type WeightInfo = weights::module_homa_lite::WeightInfo<Runtime>;
+	type Currency = Currencies;
+	type StakingCurrencyId = KSMCurrencyId;
+	type LiquidCurrencyId = LKSMCurrencyId;
+	type PalletId = HomaLitePalletId;
```
Comment (shaunxw) [inline runtime/karura/src/lib.rs:1423]: The `PalletId` is not actually used in the module now no?

File: runtime/karura/src/lib.rs [hunk 10]
```diff
@@ -1403,6 +1405,29 @@ impl cumulus_pallet_dmp_queue::Config for Runtime {
 	type ExecuteOverweightOrigin = EnsureRoot<AccountId>;
 }
 
+parameter_types! {
+	pub const KSMCurrencyId: CurrencyId = CurrencyId::Token(TokenSymbol::KSM);
+	pub const LKSMCurrencyId: CurrencyId = CurrencyId::Token(TokenSymbol::LKSM);
+	pub const MinimumMintThreshold: Balance = 1_000_000_000;
+	pub RelaychainSovereignSubAccount: MultiLocation = MultiLocation::X2( Junction::Parent, Junction::AccountId32 {
+		network: RelayNetwork::get(),
+		id: ParachainInfo::get().into_account(),
```
Comment (shaunxw) [inline runtime/karura/src/lib.rs:1414]: @xlc we may need a derive path to get the sovereign sub-account id, any suggestions?

Comment (xlc) [reply runtime/karura/src/lib.rs:1428]: https://github.com/paritytech/substrate/blob/7960c9eba8c1a73e8792c200594544d65b1333bd/frame/utility/src/lib.rs#L310

this is public, so we should be able to call it to calculate the account

File: runtime/mandala/src/lib.rs [hunk 11]
```diff
@@ -1211,18 +1210,25 @@ impl module_homa::Config for Runtime {
...
 
 parameter_types! {
-	pub const KSM: CurrencyId = CurrencyId::Token(TokenSymbol::KSM);
-	pub const LKSM: CurrencyId = CurrencyId::Token(TokenSymbol::LKSM);
+	pub const DotCurrencyId: CurrencyId = CurrencyId::Token(TokenSymbol::DOT);
+	pub const LdotCurrencyId: CurrencyId = CurrencyId::Token(TokenSymbol::LDOT);
+	pub const MinimumMintThreshold: Balance = 1_000_000_000;
+	pub RelaychainSovereignSubAccount: MultiLocation = MultiLocation::X2( Junction::Parent, Junction::AccountId32 {
+		network: RelayNetwork::get(),
+		id: Utility::derivative_account_id(ParachainInfo::get().into_account(), RELAYCHAIN_SUB_ACCOUNT_ID).into(),
```
Comment (syan095) [inline runtime/mandala/src/lib.rs:1218]: Is this the right way?

Comment (xlc) [reply runtime/mandala/src/lib.rs:1218]: looks right. can you have a test to assert this value?

Comment (shaunxw) [reply runtime/mandala/src/lib.rs:1218]: We need to asset the derived account id is different from original one, and can be used with the origin one.

Comment (syan095) [reply runtime/mandala/src/lib.rs:1218]: Added an integration test to the Mandala runtime to assert the derivative_account_id() produces unique addresses

File: runtime/mandala/tests/integration_test.rs [hunk 12]
```diff
@@ -1855,3 +1855,17 @@ fn currency_id_convert() {
...
+
+#[test]
+fn parachain_subaccounts_are_unique() {
+	ExtBuilder::default().build().execute_with(|| {
+		let parachain: AccountId = ParachainInfo::parachain_id().into_account();
+		let relaychain = Utility::derivative_account_id(ParachainInfo::get().into_account(), 0);
+		let relaychain_subaccount_1 = Utility::derivative_account_id(ParachainInfo::get().into_account(), 1);
+		let relaychain_subaccount_2 = Utility::derivative_account_id(ParachainInfo::get().into_account(), 2);
+		assert_ne!(parachain, relaychain);
+		assert_ne!(parachain, relaychain_subaccount_1);
```
Comment (xlc) [inline runtime/mandala/tests/integration_test.rs:1867]: no. I want to assert something like `relaychain_subaccount_1 == hex_literal!("12345678901234567890")`

File: runtime/mandala/tests/integration_test.rs [hunk 13]
```diff
@@ -1855,3 +1855,17 @@ fn currency_id_convert() {
...
+
+#[test]
+fn parachain_subaccounts_are_unique() {
+	ExtBuilder::default().build().execute_with(|| {
+		let parachain: AccountId = ParachainInfo::parachain_id().into_account();
+		let relaychain: AccountId = Utility::derivative_account_id(ParachainInfo::get().into_account(), 0);
+		let account1: AccountId =
+			hex_literal::hex!["7061726164000000000000000000000000000000000000000000000000000000"].into();
+		let account2: AccountId =
+			hex_literal::hex!["00465d6ab005c2fd8c4e0bf22a60fe3ce5ff035072ec74679f4babb4c6f00833"].into();
```
Comment (syan095) [inline runtime/mandala/tests/integration_test.rs:1867]: @xlc is this what you mean?

Comment (xlc) [reply runtime/mandala/tests/integration_test.rs:1867]: Yes. Just need to change the name `relaychain` to `sub_account`

File: runtime/karura/src/lib.rs [hunk 14]
```diff
@@ -1417,6 +1418,28 @@ impl cumulus_pallet_dmp_queue::Config for Runtime {
 	type ExecuteOverweightOrigin = EnsureRoot<AccountId>;
 }
 
+parameter_types! {
+	pub const KSMCurrencyId: CurrencyId = CurrencyId::Token(TokenSymbol::KSM);
+	pub const LKSMCurrencyId: CurrencyId = CurrencyId::Token(TokenSymbol::LKSM);
+	pub const MinimumMintThreshold: Balance = 1_000_000_000;
```
Comment (xlc) [inline runtime/karura/src/lib.rs:1424]:
```suggestion
	pub const MinimumMintThreshold: Balance = 10 * cent(KAR);
```

File: runtime/karura/src/lib.rs [hunk 15]
```diff
@@ -1417,6 +1418,28 @@ impl cumulus_pallet_dmp_queue::Config for Runtime {
...
+	pub RelaychainSovereignSubAccount: MultiLocation = MultiLocation::X2( Junction::Parent, Junction::AccountId32 {
+		network: RelayNetwork::get(),
+		id: Utility::derivative_account_id(ParachainInfo::get().into_account(), RELAYCHAIN_SUB_ACCOUNT_ID).into(),
```
Comment (xlc) [inline runtime/karura/src/lib.rs:1427]: make this a helper method and update the test to assert result of this method

**Review by xlc — APPROVED** [review]

**Comment by xlc** [comment]
Hold off merge this until 1.2.2 release is out.

**Review by shaunxw — APPROVED** [review]

**Pull request merged**